### DISCUSSION TITLE
How to link setterm -powerdown to backlight?

### DISCUSSION THREAD

**Discussion by Malvineous** [notro/fbtft#605] [Q&A]
Thanks so much for this excellent project, and for taking the time to get it merged into the kernel!

I have a Linux framebuffer console (i.e. no X11) using fb_ili9341 and everything is working fine.

I can switch the backlight off using `echo 1 > /sys/class/backlight/fb_ili9341/bl_power` (and `echo 0` to turn it back on again).

However even if I run `setterm -powerdown 1` to tell the console to switch off after one minute, nothing happens.  I can run `setterm -blank force` and the screen will immediately blank (go black) but the backlight remains on (which is normal for `-blank` but shows the `setterm` command is working correctly).

Is there a way to link in the backlight control with the console's existing power saving functionality?

**Comment by notro** [comment]
I have chased down the call chain:

[setterm](https://github.com/util-linux/util-linux/blob/master/term-utils/setterm.c) -powerdown does this: `printf("\033[14;%d]", ctl->opt_pd_min);`

In the kernel:

The [setterm_command()](https://elixir.bootlin.com/linux/latest/C/ident/setterm_command) for 14 does this: `vesa_off_interval = min(vc->vc_par[1], 60U) * 60 * HZ;`

[vesa_off_interval](https://elixir.bootlin.com/linux/latest/C/ident/vesa_off_interval) is used in [do_blank_screen](https://elixir.bootlin.com/linux/latest/C/ident/do_blank_screen) where it calls the con_blank callback.

For fbdev this callback is [fbcon_blank()](https://elixir.bootlin.com/linux/latest/C/ident/fbcon_blank) which calls [fb_blank()](https://elixir.bootlin.com/linux/latest/C/ident/fb_blank) which in turn calls the fb_blank callback [fbtft_fb_blank](https://elixir.bootlin.com/linux/latest/C/ident/fbtft_fb_blank).
This function mimicks the fbdev core behaviour and returns -EINVAL if there's no fbtft blank callback. fb_ili9341 has no such callback since it has no direct display controller control of the backlight. fb_blank() thus get -EINVAL from the callback and this line is not executed: `fb_notifier_call_chain(FB_EVENT_BLANK, &event);`

So the backlight subsystem is not told to turn off the backlight, which would have happened in [fb_notifier_callback()](https://elixir.bootlin.com/linux/latest/C/ident/fb_notifier_callback).

I think the solution to this problem is to have fbtft_fb_blank() return zero even if the blank callback is not set:

```
 static int fbtft_fb_blank(int blank, struct fb_info *info)
 {
 	struct fbtft_par *par = info->par;
- 	int ret = -EINVAL;
+ 	int ret = 0;
 
 	dev_dbg(info->dev, "%s(blank=%d)\n",
		__func__, blank);
```

Ufortunately I haven't got the time to test this nor to send a patch to fix it.

**Reply by Malvineous** [reply]
Wow that's really informative, thanks so much for the detail!  I will see if I can do some testing.

When you say fb_ili9341 has no direct control of the backlight, what code is controlling the backlight in response to `/sys/class/backlight/fb_ili9341/bl_power`?  Is this a separate module or a different part of the code that `fbtft_fb_blank()` cannot see?

**Comment by notro** [comment]
fbtft registers a backlight device if the 'led' gpio is set. This happens in [fbtft_register_backlight](https://elixir.bootlin.com/linux/latest/C/ident/fbtft_register_backlight). Userspace controls this device by writing to the file you mentioned. 

Call chain: [bl_power_store()](https://elixir.bootlin.com/linux/latest/C/ident/bl_power_store) -> [backlight_update_status()](https://elixir.bootlin.com/linux/latest/C/ident/backlight_update_status) -> [fbtft_backlight_update_status()](https://elixir.bootlin.com/linux/latest/C/ident/fbtft_backlight_update_status)

fbdev controls the backlight subsystem using the [FB_EVENT_BLANK](https://elixir.bootlin.com/linux/latest/C/ident/FB_EVENT_BLANK) notifier event.
